File: javascript/deriv.js

function many(e) {
	return {"Type": "ZeroOrMore", "Expr": e}
}

function concat(e1, e2) {
	return {"Type": "Concat", "Expr": e1, "Expr2": e2}
}

function or(e1, e2) {
	return {"Type": "Or", "Expr": e1, "Expr2": e2}
}

function value(v) {
	return {"Type": "Value", "Value": v}
}

function ref(v) {
	return {"Type": "Reference", "Name": v}
}

function emptyset() {
	return {"Type": "Value", "Value": "emptyset"}
} 

function empty() {
	return {"Type": "Value", "Value": "empty"}
}

function nullable(refs, expr) {
	switch (expr.Type) {
	case "Value":
		switch (expr.Value) {
		case "empty":
			return true
		case "emptyset":
			return false
		}
		return false
	case "Or":
		return nullable(refs, expr.Expr) || nullable(refs, expr.Expr2)
	case "Concat":
		return nullable(refs, expr.Expr) && nullable(refs, expr.Expr2)
	case "ZeroOrMore":
		return true
	case "Reference":
		return nullable(refs, refs[expr.Name])
	default:
		console.log("unknown type: " + expr.Type)
	}
}

function toStr(expr) {
	switch (expr.Type) {
	case "Value":
		return "\"" + expr.Value + "\""
	case "Or":
		return "or(" + toStr(expr.Expr) + ", " + toStr(expr.Expr2) + ")"
	case "Concat":
		return "concat(" + toStr(expr.Expr) + ", " + toStr(expr.Expr2) + ")"
	case "ZeroOrMore":
		return "many(" + toStr(expr.Expr) + ")"
	case "Reference":
		return "ref(" + expr.Name + ")"
	default:
		throw("unknown type: " + JSON.stringify(expr))
	}
}

function isEmpty(expr) {
	return (expr.Type == "Value") && (expr.Value == "empty")
}

function isEmptySet(expr) {
	return (expr.Type == "Value") && (expr.Value == "emptyset")
}

function simplify(expr) {
	switch (expr.Type) {
	case "Value":
		return expr
	case "Or":
		var left = simplify(expr.Expr);
		var right = simplify(expr.Expr2);
		if (isEmptySet(left)) {
			return right
		}
		if (isEmptySet(right)) {
			return left
		}
		if (JSON.stringify(left) == JSON.stringify(right)) {
			return left
		}
		return or(left, right)
	case "Concat":
		var left = simplify(expr.Expr);
		var right = simplify(expr.Expr2);
		if (isEmpty(left)) {
			return right
		}
		if (isEmptySet(left)) {
			return emptyset()
		}
		if (isEmptySet(right)) {
			return emptyset()
		}
		return concat(left, right)
	case "ZeroOrMore":
		return many(simplify(expr.Expr))
	case "Reference":
		return expr
	default:
		throw("unknown type: " + JSON.stringify(expr))
	}
}

function satis(refs, expr) {
	switch (expr.Type) {
	case "Value":
		switch (expr.Value) {
		case "empty":
			return true
		case "emptyset":
			return false
		}
		return true
	case "Or":
		return satis(refs, expr.Expr) || satis(refs, expr.Expr2)
	case "Concat":
		return satis(refs, expr.Expr) && satis(refs, expr.Expr2)
	case "ZeroOrMore":
		return true
	case "Reference":
		return satis(refs, refs[expr.Name])
	default:
		throw("unknown type: " + JSON.stringify(expr))
	}
}

function deriv(refs, expr, c) {
	switch (expr.Type) {
	case "Value":
		switch (expr.Value) {
		case "empty":
			return emptyset()
		case "emptyset":
			return emptyset()
		}
		if (expr.Value == c) {
			return empty()
		} else {
			return emptyset()
		}
	case "Or":
		var left = deriv(refs, expr.Expr, c);
		var right = deriv(refs, expr.Expr2, c);
		return or(left, right)
	case "Concat":
		if (nullable(refs, expr.Expr)) {
			var left = deriv(refs, expr.Expr, c);
			var right = deriv(refs, expr.Expr2, c);
			return or(concat(left, expr.Expr2), right)
		} else {
			var left = deriv(refs, expr.Expr, c);
			return concat(left, expr.Expr2)
		}
	case "ZeroOrMore":
		var e = deriv(refs, expr.Expr, c);
		return concat(e, expr)
	case "Reference":
		return deriv(refs, refs[expr.Name], c)
	default:
		throw("unknown type: " + JSON.stringify(expr))
	}
}

File: golang/deriv/deriv.go
package deriv

import (
	"fmt"
	. "github.com/awalterschulze/ragax/golang/ast"
	"github.com/awalterschulze/ragax/golang/lexer"
	"github.com/awalterschulze/ragax/golang/parser"
)

func Parse(s string) (Defs, error) {
	p := parser.NewParser()
	l := lexer.NewLexer([]byte(s))
	r, err := p.Parse(l)
	if err != nil {
		return nil, err
	}
	return r.(Defs), nil
}

func Eval(defs Defs, s string) *Expr {
	d := defs[0].Expr.Expr()
	mem := make(Mem)
	deriv := &deriv{mem: mem, defs: defs}
	for _, c := range s {
		fmt.Printf("-- %s deriv %s\n", d, string([]rune{c}))
		d = deriv.Deriv(d, c)
		fmt.Printf("== res %s\n", d)
	}
	return d
}

var emptyset = &Expr{Type: Value, Value: "emptyset"}
var empty = &Expr{Type: Value, Value: "empty"}

type Mem map[*Expr]map[rune]*Expr

func (m Mem) Get(e *Expr, r rune) *Expr {
	mm, ok := m[e]
	if ok {
		return mm[r]
	}
	return nil
}

func (m Mem) Set(e *Expr, r rune, v *Expr) *Expr {
	if _, ok := m[e]; !ok {
		m[e] = make(map[rune]*Expr)
	}
	m[e][r] = v
	return v
}

type deriv struct {
	mem  Mem
	defs Defs
}

func (this *deriv) Sderiv(e *Expr, c rune) *Expr {
	return Simplify(this.defs, this.Deriv(e, c))
}

func (this *deriv) Deriv(e *Expr, c rune) *Expr {
	if res := this.mem.Get(e, c); res != nil {
		return res
	}
	fmt.Printf("%s deriv %v\n", e, string([]rune{c}))
	switch e.Type {
	case Value:
		switch e.Value {
		case "empty":
			return this.mem.Set(e, c, emptyset)
		case "emptyset":
			return this.mem.Set(e, c, emptyset)
		default:
			if e.Value == string([]rune{c}) {
				return this.mem.Set(e, c, empty)
			}
			return this.mem.Set(e, c, emptyset)
		}
	case Or:
		return this.mem.Set(e, c, &Expr{
			Type: Or,
			Expr: NewLazy(func() *Expr {
				return this.Sderiv(e.Expr.Expr(), c)
			}),
			Expr2: NewLazy(func() *Expr {
				return this.Sderiv(e.Expr2.Expr(), c)
			}),
		})
	case Concat:
		if Nullable(this.defs, e.Expr) {
			return this.mem.Set(e, c, &Expr{
				Type: Or,
				Expr: NewExpr(&Expr{
					Type: Concat,
					Expr: NewLazy(func() *Expr {
						return this.Sderiv(e.Expr.Expr(), c)
					}),
					Expr2: e.Expr2,
				}),
				Expr2: NewLazy(func() *Expr {
					return this.Sderiv(e.Expr2.Expr(), c)
				}),
			})
		} else {
			return this.mem.Set(e, c, &Expr{
				Type: Concat,
				Expr: NewLazy(func() *Expr {
					return this.Sderiv(e.Expr.Expr(), c)
				}),
				Expr2: e.Expr2,
			})
		}
	case ZeroOrMore:
		return this.mem.Set(e, c, &Expr{
			Type: Concat,
			Expr: NewLazy(func() *Expr {
				return this.Sderiv(e.Expr.Expr(), c)
			}),
			Expr2: NewExpr(Simplify(this.defs, &Expr{Type: ZeroOrMore, Expr: e.Expr})),
		})
	case Reference:
		expr1 := lookup(this.defs, e.Value).Expr()
		return this.mem.Set(e, c, this.Deriv(expr1, c))
	}
	panic("unreachable")
}

func lookup(defs Defs, name string) *LazyExpr {
	for _, def := range defs {
		if def.Name == name {
			return def.Expr
		}
	}
	return nil
}

type nullable struct {
	defs    Defs
	mem     map[*LazyExpr]bool
	visited map[*LazyExpr]bool
}

func (this *nullable) Nullable(l *LazyExpr) bool {
	if m, ok := this.mem[l]; ok {
		return m
	}
	if v := this.visited[l]; v {
		this.mem[l] = false
		return false
	}
	this.visited[l] = true
	e := l.Expr()
	switch e.Type {
	case Value:
		switch e.Value {
		case "empty":
			this.mem[l] = true
			return true
		case "emptyset":
			this.mem[l] = false
			return false
		default:
			this.mem[l] = false
			return false
		}
	case Or:
		v := this.Nullable(e.Expr) || this.Nullable(e.Expr2)
		this.mem[l] = v
		return v
	case Concat:
		v := this.Nullable(e.Expr) && this.Nullable(e.Expr2)
		this.mem[l] = v
		return v
	case ZeroOrMore:
		this.mem[l] = true
		return true
	case Reference:
		v := this.Nullable(lookup(this.defs, e.Value))
		this.mem[l] = v
		return v
	}
	panic(fmt.Sprintf("unreachable %#v", e))
}

func Nullable(defs Defs, l *LazyExpr) bool {
	n := &nullable{
		defs:    defs,
		mem:     make(map[*LazyExpr]bool),
		visited: make(map[*LazyExpr]bool),
	}
	return n.Nullable(l)
}

func isEmptySet(e *Expr) bool {
	return e.Type == Value && e.Value == "emptyset"
}

func isEmpty(e *Expr) bool {
	return e.Type == Value && e.Value == "empty"
}

func Simplify(defs Defs, e *Expr) *Expr {
	switch e.Type {
	case Value:
		return e
	case Or:
		if e.Expr.IsLazy() || e.Expr2.IsLazy() {
			return e
		}
		expr1 := Simplify(defs, e.Expr.Expr())
		expr2 := Simplify(defs, e.Expr2.Expr())
		if isEmptySet(expr1) {
			return expr2
		}
		if isEmptySet(expr2) {
			return expr1
		}
		return &Expr{Type: Or, Expr: NewExpr(expr1), Expr2: NewExpr(expr2)}
	case Concat:
		if e.Expr.IsLazy() || e.Expr2.IsLazy() {
			return e
		}
		expr1 := Simplify(defs, e.Expr.Expr())
		expr2 := Simplify(defs, e.Expr2.Expr())
		if isEmptySet(expr1) {
			return emptyset
		}
		if isEmpty(expr1) {
			return expr2
		}
		if isEmptySet(expr2) {
			return emptyset
		}
		return &Expr{Type: Concat, Expr: NewExpr(expr1), Expr2: NewExpr(expr2)}
	case ZeroOrMore:
		if e.Expr.IsLazy() {
			return e
		}
		expr1 := Simplify(defs, e.Expr.Expr())
		if isEmptySet(expr1) {
			return empty
		}
		return &Expr{Type: ZeroOrMore, Expr: NewExpr(expr1)}
	case Reference:
		return e
	}
	panic(fmt.Sprintf("unreachable %#v", e))
}
